{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "819a2208",
   "metadata": {},
   "outputs": [],
   "source": [
    "# An attempt to model RNA evolution using a large language model (LLM)\n",
    "# In config/rna_rag.yaml, set prompt_choice: 1 to generate RNA sequences\n",
    "# go to cd drfold2 and start drfold_container\n",
    "# docker run --gpus all -it --name drfold_container -v .:/opt/drfold2 drfold_image bash\n",
    "\n",
    "# Load configuration file and check for appropriate prompt\n",
    "import yaml\n",
    "with open(\"../configs/rna_rag.yaml\", 'r') as f:\n",
    "    global_config = yaml.safe_load(f)\n",
    "assert(global_config[\"llm_model\"][\"prompt_choice\"]==1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55714952",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import time\n",
    "sys.path.append(os.path.abspath('../'))\n",
    "sys.path.append(os.path.abspath('../RiboDiffusion'))\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from PIL import Image\n",
    "\n",
    "from lib.fasta_utils import read_single_fasta, write_single_fasta\n",
    "\n",
    "from importlib import reload\n",
    "\n",
    "import lib.lib_drfold\n",
    "reload(lib.lib_drfold)  # Reload the module\n",
    "from lib.lib_drfold import clear_output, fasta2pdb\n",
    "\n",
    "import lib.lib_ribodiffusion\n",
    "reload(lib.lib_ribodiffusion)  # Reload the module\n",
    "from lib.lib_ribodiffusion import Pdb2Fasta\n",
    "pdb2fasta = Pdb2Fasta()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc46c0ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "# RAG class instantiation\n",
    "\n",
    "from importlib import reload\n",
    "from lib.RAG_Biomistral.RAG_Biomistral import RAG_Biomistral\n",
    "reload(lib.RAG_Biomistral.RAG_Biomistral)  # Reload the module\n",
    "from lib.RAG_Biomistral.RAG_Biomistral import RAG_Biomistral\n",
    "\n",
    "rag = RAG_Biomistral()\n",
    "\n",
    "# Calculate embeddings for RNA books\n",
    "#path2rnaBooks =  [\"../data/raw/books/mattick.pdf\",] # pdfs are here\n",
    "path2rnaBooks = [os.path.join(r, f) for r, _, files in os.walk(\"../data/raw/books\") for f in files if f.lower().endswith(\".pdf\")]\n",
    "if len(path2rnaBooks)==0:\n",
    "    raise FileNotFoundError(\"No PDF books found. Please add PDF files to ../data/raw/books.\")\n",
    "path2bookEmbed = '../data/processed/RAG/book_embeddings.pkl' # Save embeddings to save time on future runs\n",
    "rag.calc_book_embeddings(path2rnaBooks=path2rnaBooks, path2bookEmbed=path2bookEmbed, save=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5664b244",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define folders, files, RAG/LLM initialization\n",
    "\n",
    "allowed_nucleotides = {'A', 'U', 'G', 'C'}\n",
    "\n",
    "input_fasta_dir = \"../data/processed/RAG/rag_pipeline/\"\n",
    "output_dir = \"../data/processed/RAG/RNA_evolution/\"\n",
    "os.makedirs(output_dir, exist_ok=True)\n",
    "\n",
    "fasta_files = [\"trna_glycine\"]\n",
    "num_evolution_steps = 200\n",
    "for i in range(num_evolution_steps):\n",
    "    fasta_files.append(str(i))\n",
    "\n",
    "sequence_0 = read_single_fasta(input_fasta_dir+fasta_files[0]+\".fasta\")[1]\n",
    "print(\"Initial sequence\", sequence_0)\n",
    "clear_output(output_dir)\n",
    "write_single_fasta(header=\"seq_\"+str(len(sequence_0))+\"nt\", sequence=sequence_0, filepath=output_dir+fasta_files[0]+\".fasta\")\n",
    "\n",
    "rag.preprocess() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2b93d9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loop through evolution steps\n",
    "# Output saved to: ../data/processed/RAG/RNA_evolution/\n",
    "\n",
    "import lib.lib_x3dna\n",
    "reload(lib.lib_x3dna)  # Reload the module\n",
    "from lib.lib_x3dna import RnaProperties\n",
    "\n",
    "for i in range(len(fasta_files)-1):\n",
    "\n",
    "    print(\"-----> STEP\", i)\n",
    "    time.sleep(1)\n",
    "\n",
    "    print(\"- calculate PDB from fasta\")\n",
    "    fasta_file = output_dir + fasta_files[i] + \".fasta\"\n",
    "    sequence = read_single_fasta(fasta_file)[1]\n",
    "    fasta2pdb(source_path = fasta_file)\n",
    "    pdb_file = fasta_file.replace('.fasta', '.pdb')\n",
    "\n",
    "    print()\n",
    "    print(\"- calculating PDB features...\")\n",
    "    fromPDB = RnaProperties(sequence=sequence, pdb_input=pdb_file)\n",
    "    PDBprompt = {}\n",
    "    PDBprompt[\"general_info\"] = fromPDB.get_genInfo()\n",
    "    PDBprompt[\"helices_info\"] = fromPDB.get_helicesInfo() # \"\" if no helices\n",
    "    # GET and save 3D structure image\n",
    "    pdbImage = fromPDB.get_pdbImage()\n",
    "    image_file = pdb_file.replace('.pdb', '.png')\n",
    "    plt.imsave(image_file, pdbImage, cmap='viridis')\n",
    "    rag_images = [Image.open(image_file).convert('RGB')]\n",
    "\n",
    "    print()\n",
    "    print(\"- calculating backfolded_rna\")\n",
    "    output = pdb2fasta.pdb2fasta(source_path = pdb_file, cond_scale=-1.)\n",
    "    output_sorted = sorted(output, key=lambda x: x[1], reverse=True) # sort by similarity to main_sequence\n",
    "\n",
    "    print()\n",
    "    print(\"- LLM part...\")\n",
    "    index_backfolded_rna = 0\n",
    "    while index_backfolded_rna<len(output_sorted):\n",
    "        # Generate the ancestral RNA sequence\n",
    "        backfolded_rna_tuple = output_sorted[index_backfolded_rna]\n",
    "        print(\"XXXXXX sequence:\", backfolded_rna_tuple[0], backfolded_rna_tuple[1])\n",
    "        answer = rag.process(main_sequence=sequence, backfolded_rna_tuple=backfolded_rna_tuple, \n",
    "            PDBprompt=PDBprompt, rag_images=rag_images)\n",
    "        print(sequence)\n",
    "        sequence_llm = answer.split(\"\\n\")[-1].strip().split(\" \")[-1].replace('\"', '').replace(':', '')\n",
    "\n",
    "        # Check whether the ancestral RNA sequence is correct\n",
    "        true_sequence = all(nucleotide in allowed_nucleotides for nucleotide in sequence_llm)\n",
    "        if (sequence_llm==sequence) or (not true_sequence) or (sequence_llm==\"\"):\n",
    "            index_backfolded_rna +=1\n",
    "            print(\"Problems with RNA generation by llm. Trying again with index_backfolded_rna = {}\".format(index_backfolded_rna))\n",
    "        else: # (sequence_llm != sequence) and true_sequence\n",
    "            print(\"Valid RNA sequence, generated by LLM, on attempt {}\".format(index_backfolded_rna))\n",
    "            sequence = sequence_llm\n",
    "            break     \n",
    "    print(sequence)\n",
    "\n",
    "    fasta_file_next = output_dir + fasta_files[i+1] + \".fasta\"\n",
    "    write_single_fasta(header=\"seq\", sequence=sequence, filepath=fasta_file_next)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac4d12fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compare two RNA sequences\n",
    "\n",
    "from Bio.Align import PairwiseAligner\n",
    "\n",
    "# Example RNA sequences\n",
    "rna1 = \"GGGGCCAUAGCUCAAUUGGCAGAGCGCCGCCUUUGCAAGGCGGAGGCUAGGGGUUCGAUUCCCCUUGGCUCCA\" \n",
    "rna2 = \"GGGGCCAUAGCUCAAUUGGCAGAGCGCCGCCUUUGCAAGGUAGAGGUCCGGGGUUCGAAUCCCCGCGCCUCCC\" \n",
    "\n",
    "\n",
    "\n",
    "# Create and Configure the Aligner\n",
    "aligner = PairwiseAligner()\n",
    "\n",
    "# Optional: adjust scoring\n",
    "aligner.match_score = 2\n",
    "aligner.mismatch_score = -1\n",
    "aligner.open_gap_score = -2\n",
    "aligner.extend_gap_score = -0.5\n",
    "\n",
    "# Run the Alignment\n",
    "alignments = aligner.align(rna1, rna2)\n",
    "\n",
    "# Print the Best Alignment\n",
    "best_alignment = alignments[0]\n",
    "print(type(best_alignment))\n",
    "print(best_alignment)\n",
    "print(f\"Score: {best_alignment.score}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "rna_rag2",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
